thesis_text/thesis_text.md

```diff
@@ -0,0 +1,551 @@
+
+To fully appreciate the importance of this strategy, we have to point out the following fact of CLI's design from _I.12.3.2.1 The evaluation stack_  (emphasis added):
+
+> The type state of the stack (**the stack depth** and types of each element on the stack) at any given point in a program **shall be identical for all possible control flow paths**. For example, a program that loops an unknown number of times and pushes a new element on the stack at each iteration would be prohibited. 
+
+This design choice is not a coincidence, as it is vital also for hand-crafting performant JIT compilers. With regards to `MERGE_EXPLODE`, it means that all copies of the interpreter's inner loop that have the same bytecode offset will also have the same evaluation stack depth and type layout.
+
+Thanks to this, if we have for example a push immediate 4 instruction somewhere in the code, we can be sure it can be translated to a simple statement like `stack[7] = 4`, as in every execution of this instruction the stack depth has to be the same. This enables more optimizations, as this constant can be propagated to the next instruction reading `stack[7]`. 
+
+For a more involved example, let's apply this strategy to the following pseudo bytecode of `for(int i = 0; i < 100; i++) {a = a*a; }; return a;`:
+
+```
+; i=0
+0: OPCODE_LOADCONST 0
+1: OPCODE_STOREVAR i
+
+; i < 100
+2: OPCODE_LOADVAR i
+3: OPCODE_LOADCONST 100
+4: OPCODE_JMPIFBEQ @14
+
+; a = a*a
+5: OPCODE_LOADVAR a
+6: OPCODE_LOADVAR a
+7: OPCODE_MULTIPLY
+8: OPCODE_STOREVAR a
+
+; i++
+9: OPCODE_LOADVAR i
+10: OPCODE_LOADCONST 1
+11: OPCODE_ADD
+12: OPCODE_STOREVAR i
+
+; loop
+13: OPCODE_JMP @2
+
+; return a
+14: OPCODE_LOADVAR a
+15: OPCODE_RET
+```
+
+Thanks to the strategy, for every bytecode offset only one state has to be created. Knowing exactly the stack depth, we can partially evaluate the stack positions to constants
+
+```
+; i=0
+0: stack[0] = 0
+1: vars[i] = stack[0]
+
+; i < 100
+2: stack[0] = vars[i]
+3: stack[1] = 100
+4: if(stack[0]>=stack[1]) goto @14
+
+; a = a*a
+5: stack[0] = vars[a]
+6: stack[1] = vars[a]
+7: stack[0] = stack[0] * stack[1]
+8: vars[a] = stack[0]
+
+; i++
+9: stack[0] = vars[i]
+10: stack[1] = 1
+11: stack[0] = stack[0] + stack[1]
+12: vars[i] = stack[0]
+
+; loop
+13: goto @2
+
+; return a
+14: stack[0] = vars[a]
+15: return stack[0]
+```
+
+As the stack is not used outside this method, it will be completely virtualized. The fact that all stack array accesses use constant allows for aggressive optimization resulting in completely optimizing out the array:
+
+```
+  vars[i] = 0;
+
+condition:
+  if(vars[i]>=100) goto end
+  vars[a] = vars[a]*vars[a]
+  vars[i] = vars[i]+1
+  goto condition
+
+end:
+  return vars[a]
+```
+
+The original control flow of the method was reconstructed from flat bytecode just by unrolling the interpreter loop and merging the instances having the same bytecode offset.
```

Comment:
Neni mi to ted jasne: "was reconstructed from flat bytecode just by unrolling the interpreter loop" -- jakoze jsme si to ted rucne udelali, abychom pak chapali, co dela Graal/Truffle compiler? Nebo was reconstructed by Truffle compiler? V tom pripade by tu mohl byt skec toho interpreter loopu?